File: org.eclipse.scout.rt.ui.html/src/main/js/scout/util/Range.js
/*******************************************************************************
 * Copyright (c) 2014-2016 BSI Business Systems Integration AG.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     BSI Business Systems Integration AG - initial API and implementation
 ******************************************************************************/
scout.Range = function(from, to) {
  this.from = from;
  this.to = to;
};

scout.Range.prototype.equals = function(other) {
  return this.from === other.from && this.to === other.to;
};

scout.Range.prototype.subtract = function(other) {
  // other is empty
  if (other.size() === 0) {
    return [new scout.Range(this.from, this.to)];
  }
  // other is greater
  if (this.from >= other.from && this.to <= other.to) {
    return [new scout.Range(0, 0)];
  }
  // other is contained completely
  if (other.from >= this.from && other.to <= this.to) {
    var range1 = new scout.Range(this.from, other.from);
    var range2 = new scout.Range(other.to, this.to);
    if (range1.size() === 0) {
      return [range2];
    }
    if (range2.size() === 0) {
      return [range1];
    }
    return [range1, range2];
  }
  // other overlaps on the bottom
  if (other.from > this.from && other.from < this.to) {
    return [new scout.Range(this.from, other.from)];
  }
  // other overlaps on the top
  if (this.from > other.from && this.from < other.to) {
    return [new scout.Range(other.to, this.to)];
  }
  // other is outside
  return [new scout.Range(this.from, this.to)];
};


scout.Range.prototype.shrink = function(other) {
  // other is empty
  if (other.size() === 0) {
    return new scout.Range(this.from, this.to);
  }
  // other is greater
  if (this.from >= other.from && this.to <= other.to) {
    return new scout.Range(0, 0);
  }
  // other is contained completely
  if (other.from >= this.from && other.to <= this.to) {
    return new scout.Range(this.from, other.to);
  }
  // other overlaps on the bottom
  if (other.from >= this.from && other.from < this.to) {
    return new scout.Range(this.from, other.from);
  }
  // other overlaps on the top
  if (this.from > other.from && this.from < other.to) {
    return new scout.Range(other.to, this.to);
  }
  if(other.to<this.from){
    return new scout.Range(this.from-other.size()-1, this.to-other.size()-1);
  }
  // other is outside
  return new scout.Range(this.from, this.to);
};


scout.Range.prototype.union = function(other) {
  if (this.to < other.from || other.to < this.from) {
    var range1 = new scout.Range(this.from, this.to);
    var range2 = new scout.Range(other.from, other.to);
    if (range1.size() === 0) {
      return [range2];
    }
    if (range2.size() === 0) {
      return [range1];
    }
    return [range1, range2];
  }
  return [new scout.Range(Math.min(this.from, other.from), Math.max(this.to, other.to))];
};

scout.Range.prototype.intersect = function(other) {
  if (this.to <= other.from || other.to <= this.from) {
    return new scout.Range(0, 0);
  }
  return new scout.Range(Math.max(this.from, other.from), Math.min(this.to, other.to));
};

scout.Range.prototype.size = function(other) {
  return this.to - this.from;
};

scout.Range.prototype.contains = function(value) {
  return this.from <= value && value < this.to;
};

scout.Range.prototype.toString = function() {
  return 'scout.Range[' +
    'from=' + (this.from === null ? 'null' : this.from) +
    ' to=' + (this.to === null ? 'null' : this.to) + ']';
};

File: org.eclipse.scout.rt.platform/src/main/java/org/eclipse/scout/rt/platform/util/Range.java
/*
 * Copyright (c) 2010, 2023 BSI Business Systems Integration AG
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.scout.rt.platform.util;

import java.io.Serializable;

public class Range<VALUE_TYPE> implements Serializable {

  private static final long serialVersionUID = 1L;

  private VALUE_TYPE m_from;
  private VALUE_TYPE m_to;

  public Range() {
    this(null, null);
  }

  public Range(VALUE_TYPE from, VALUE_TYPE to) {
    m_from = from;
    m_to = to;
  }

  public Range(Range<VALUE_TYPE> rangeToCopy) {
    this(rangeToCopy != null ? rangeToCopy.getFrom() : null, rangeToCopy != null ? rangeToCopy.getTo() : null);
  }

  public void setFrom(VALUE_TYPE from) {
    m_from = from;
  }

  public VALUE_TYPE getFrom() {
    return m_from;
  }

  public void setTo(VALUE_TYPE to) {
    m_to = to;
  }

  public VALUE_TYPE getTo() {
    return m_to;
  }

  @Override
  public int hashCode() {
    final int prime = 31;
    int result = 1;
    result = prime * result + ((m_from == null) ? 0 : m_from.hashCode());
    result = prime * result + ((m_to == null) ? 0 : m_to.hashCode());
    return result;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }
    if (obj == null) {
      return false;
    }
    if (getClass() != obj.getClass()) {
      return false;
    }
    Range other = (Range) obj;
    if (m_from == null) {
      if (other.m_from != null) {
        return false;
      }
    }
    else if (!m_from.equals(other.m_from)) {
      return false;
    }
    if (m_to == null) {
      if (other.m_to != null) {
        return false;
      }
    }
    else if (!m_to.equals(other.m_to)) {
      return false;
    }
    return true;
  }

  @Override
  public String toString() {
    return "Range [m_from=" + m_from + ", m_to=" + m_to + "]";
  }

}
